src/main/java/org/mastodon/mamut/fitting/FitEllipsoidPlugin.java

```diff
@@ -187,59 +194,137 @@ private void fitSelectedVertices()
-		final double maxFactor = 1.1;
-
-		final int numSamples = 100;
-		final double outsideCutoffDistance = 3;
-		final double insideCutoffDistance = 5;
-		final double angleCutoffDistance = 30 * Math.PI / 180.0;
-		final double maxCenterDistance = 10;
-		// ----------------------------------
-
 		final MamutAppModel appModel = pluginAppModel.getAppModel();
 
 		final RefSet< Spot > vertices = appModel.getSelectionModel().getSelectedVertices();
-//		if ( vertices.isEmpty() )
-//			System.err.println( "no vertex selected" );
+		if ( vertices.isEmpty() )
+			System.err.println( "no vertex selected" );
 
 		final AffineTransform3D sourceToGlobal = new AffineTransform3D();
+		// init watch and counters
+		{
+			watch.reset();
+			watch.start();
+			found.set( 0 );
+			notFound.set( 0 );
+		}
+
+		// Fixed thread number depending on the number of available processors
+		int processors = Runtime.getRuntime().availableProcessors();
+		ExecutorService executorService = Executors.newFixedThreadPool( processors );
+		List< Callable< Object > > todo = new ArrayList<>( vertices.size() );
+
+		// parallelize over vertices
 		for ( final Spot spot : vertices )
 		{
+			Spot spotCopy = appModel.getModel().getGraph().vertexRef();
+			spotCopy.refTo( spot );
+			todo.add( Executors.callable(
+					new SpotEllipsoidFittingTask<>( spotCopy, source, sourceToGlobal, appModel, vertices.size() ) ) );
+		}
+
+		// invoke all tasks
+		try
+		{
+			executorService.invokeAll( todo );
+		}
+		catch ( InterruptedException e )
+		{
+			e.printStackTrace();
+		}
+
+		System.out.println( "found: " + found.get() + " ("
+				+ Math.round( ( double ) found.get() / ( found.get() + notFound.get() ) * 100d )
+				+ "%), not found: " + notFound.get() + " ("
+				+ Math.round( ( double ) notFound.get() / ( found.get() + notFound.get() ) * 100d )
+				+ "%), total time: " + watch.formatTime()
+				+ ", time per spot: " + ( int ) ( ( double ) watch.getTime() / ( found.get() + notFound.get() ) )
+				+ "ms." );
+	}
+
+	// TODO: move to imglib2 core and make versions for int[], double[] ???
+	public static < T extends EuclideanSpace > long[] longArrayFrom( final T t, final BiConsumer< T, long[] > get )
+	{
+		final long[] a = new long[ t.numDimensions() ];
+		get.accept( t, a );
+		return a;
+	}
+
+	private class SpotEllipsoidFittingTask< T extends RealType< T > > implements Runnable
+	{
+		private final Spot spot;
+
+		private final SourceAndConverter< T > source;
+
+		private final AffineTransform3D sourceToGlobal;
+
+		private final MamutAppModel appModel;
+
+		private final int totalTasks;
+
+		public SpotEllipsoidFittingTask(
+				final Spot spot, final SourceAndConverter< T > source, final AffineTransform3D sourceToGlobal,
+				final MamutAppModel appModel, final int totalTasks )
+		{
+			this.spot = spot;
+			this.source = source;
+			this.sourceToGlobal = sourceToGlobal;
+			this.appModel = appModel;
+			this.totalTasks = totalTasks;
+		}
```

Comment:
This class could be replaced by a method:
```java
private void fitEllipsoid(
		final Spot spot, final SourceAndConverter< T > source, final AffineTransform3D sourceToGlobal,
		final MamutAppModel appModel, final int totalTasks ) {
	// here should be all the content off the run method
}
```
which could be called via a lambda expression:
```java
todo.add( Executors.callable( () -> fitEllipsoid( spotCopy, source, sourceToGlobal, appModel, vertices.size() ) ) );
```
The result would be easier to read and understand.